iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/api.hpp

```diff
@@ -0,0 +1,120 @@
+            }                                                                                                          \
+        }                                                                                                              \
+    } while (false)
+
+/// @brief report fatal error if expr evaluates to false
+/// @note for conditions that may actually happen during correct use
+/// @param expr boolean expression that must hold
+/// @param error error object (or code)
+#define IOX_ASSERT(expr, error) IOX_REPORT_IF(!(expr), error, FATAL)
+
+//*****************************
+//* For safe mode and debugging
+//*****************************
+
+// Later on there can be variadic versions that do not need a default message but for now we leave
+// make the more general version mandatory
+
+// There are no error codes/errors required here on purpose, as it would make the use cumbersome.
+// Instead a special internal error type is used.
+// If required, a custom error option can be added but for now location should be sufficient.
+
+/// @brief in debug mode only: report fatal error if expr evaluates to false
+/// @param expr boolean expression that must hold upon entry of the function it appears in
+/// @param message message to be logged in case of violation
+#ifdef IOX_DEBUG
+#define IOX_PRECONDITION(expr, message)                                                                                \
+    do                                                                                                                 \
+    {                                                                                                                  \
+        if (requiresHandling(iox::err::PRECONDITION_VIOLATION_CODE))                                                   \
+            if (!(expr))                                                                                               \
+                forwardError(CURRENT_SOURCE_LOCATION,                                                                  \
+                             Violation(iox::err::PRECONDITION_VIOLATION_CODE),                                         \
+                             PRECONDITION_VIOLATION,                                                                   \
+                             message);                                                                                 \
+    } while (false)
+#else
+#define IOX_PRECONDITION(expr, message)                                                                                \
+    do                                                                                                                 \
+    {                                                                                                                  \
+        discard([&]() { return expr; });                                                                               \
```

Comment:
Note that this is a trick to
1. avoid evaluating the expression unless it is needed
2. avoid unused variable warnings in case the whole check is optimized out (`IOX_DEBUG` not set).

We need to consider to that `IOX_DEBUG` as default. Or call switch it around and deactivate the checks with `IOX_UNSAFE`. This is a minor issue and I am fine with both ways.

Reply:
Not required anymore, removed.